cmd/thriftlint/lint.go

```diff
@@ -0,0 +1,193 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"text/template"
+
+	"go.uber.org/thriftrw/ast"
+)
+
+// Severity of an issue
+type Severity int
+
+// Varying levels of issue severity.
+const (
+	Advice Severity = iota
+	Warning
+	Error
+)
+
+// Lint is an issue found in a Thrift file.
+type Lint struct {
+	Severity Severity
+
+	// If non-zero, this is the line in the Thrift file on which this problem
+	// was found.
+	Line int
+
+	// Message is a plain-text message to the user indicating what the problem
+	// was.
+	Message string
+}
+
+// Catcher receives lint messages from different linter rules.
+type Catcher interface {
+	PostLint(Lint)
+}
+
+// Rule is a Linter rule.
+type Rule struct {
+	// Name of the rule
+	Name string
+
+	// Inspector walks the Thrift AST and posts linter messages to the given
+	// catcher.
+	Inspector func(Catcher) ast.Visitor
+}
+
+// SimpleRule provides a convenient API to write linter rules that don't need
+// full control over the AST traversal and that post at most one message per
+// node.
+//
+// SimpleRule linters call Inspect on matching node types and for any call
+// that fails, the MessageTemplate is rendered using that Node.
+type SimpleRule struct {
+	Name string
+
+	// Severity of messages posted by this rule.
+	Severity Severity
+
+	// A function in the form,
+	//
+	// 	func(w ast.Walker, n N) bool
+	//
+	// Where N is ast.Node or any type that implements ast.Node.
+	Inspect interface{}
```

Comment:
such reflection. have you considered just writing each inspect function as an optional field?